editor/src/clj/editor/code/view.clj

```diff
@@ -3045,6 +3190,72 @@
     (.setFill gc Color/WHITE)
     (.fillText gc (format "%.3f fps" fps) (- right 5.0) (+ top 16.0))))
 
+(defn- plaintext->markdown [s]
+  (str "<p>"
+       (string/replace s #"[<>]" {"<" "&lt;" ">" "&gt;"})
+       "</p>"))
+
+(defn- handle-hover-popup-mouse-entered! [view-node _]
+  (some-> (get-property view-node :hover-request) ui/cancel))
+
+(defn- handle-hover-popup-mouse-exited! [view-node _]
+  (set-properties! view-node nil (schedule-hover-refresh! view-node)))
+
+(defn handle-hover-popup-auto-hide! [view-node _]
+  (hide-hover! view-node))
+
+(defn- hover-popup-view [props]
+  (let [{:keys [canvas-repaint-info project screen-bounds
+                hover-showing-regions hover-showing-region view-node]} props
+        {:keys [^Canvas canvas layout lines]} canvas-repaint-info
+        ^Rect r (->> hover-showing-region
+                     (data/adjust-cursor-range lines)
+                     (data/cursor-range-rects layout lines)
+                     first)
+        anchor (.localToScreen canvas
+                               (min (max 0.0 (.-x r)) (.getWidth canvas))
+                               (min (max 0.0 (.-y r)) (.getHeight canvas)))
+        ^Rectangle2D screen-rect (coll/some #(when (.contains ^Rectangle2D % anchor) %)
+                                            screen-bounds)
+        max-popup-height (- (.getY anchor) (.getMinY screen-rect))]
+    {:fx/type fxui/with-popup
+     :desc {:fx/type fxui/ext-value :value canvas}
+     :showing true
+     :anchor-x (- (.getX anchor) 4)
+     :anchor-y (+ (.getY anchor) 6)
+     :anchor-location :window-bottom-left
+     :auto-hide true
+     :on-auto-hide (fn/partial handle-hover-popup-auto-hide! view-node)
+     :auto-fix true
+     :hide-on-escape true
+     :consume-auto-hiding-events true
+     :content
+     [{:fx/type fx.stack-pane/lifecycle
+       :stylesheets [(str (io/resource "editor.css"))]
+       :on-mouse-entered (fn/partial handle-hover-popup-mouse-entered! view-node)
+       :on-mouse-exited (fn/partial handle-hover-popup-mouse-exited! view-node)
+       :children
+       [{:fx/type fx.region/lifecycle
+         :min-width 10
+         :min-height 10
+         :style-class "hover-background"}
+        {:fx/type markdown/view
+         :content (->> hover-showing-regions
+                       (mapcat
```

Comment:
Avoid the non-transducer arity of `core.mapcat` - it does a really slow lazy sequence thing internally. Prefer the transducer flavor, `coll/mapcat`, or `e/mapcat`.